---
title: "R_Assignment"
author: "Vencke"
date: "2025-03-17"
output: html_document
---

```{r load libraries, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(dplyr)
library(readr)
library(gridExtra)
```
## Part 1
### Reading in Data
All libraries were loaded in the beginning. I read in the two files, txt is tab delimited so I used read_tsv. 

```{r Read Data}
download.file("https://raw.githubusercontent.com/EEOB-BioData/BCB546_Spring2025/main/assignments/UNIX_Assignment/fang_et_al_genotypes.txt", 
              destfile = "./fang_et_al_genotypes.txt")
download.file("https://raw.githubusercontent.com/EEOB-BioData/BCB546_Spring2025/main/assignments/UNIX_Assignment/snp_position.txt", 
              destfile = "./snp_position.txt")

rm(list = ls())

snp <- read_tsv("./snp_position.txt",  show_col_types = FALSE)
fang <- read_tsv("./fang_et_al_genotypes.txt",  show_col_types = FALSE)
```

### Data Inspection
* Look at the files to see if imported correctly
* Rows, Columns and Size of Files
* Assignment says import as data frame, double check if they are data frames

```{r Data Inspection}
dim(fang)
dim(snp)

class(fang)
class(snp)

(data_inspection <- data.frame(
  Dataset = c("snp", "fang"),
  Rows = c(nrow(snp), nrow(fang)),
  Columns = c(ncol(snp), ncol(fang)),
  Size_MB = c(
    format(object.size(snp) / 1024^2, digits = 2),
    format(object.size(fang) / 1024^2, digits = 2)
  ),
  IsDataFrame = c(is.data.frame(snp), is.data.frame(fang))  
))
```

### Data Processing - filter & transpose
* Create maize and teosinte subset


```{r filter & transpose}
maize <- filter(fang, Group %in% c("ZMMIL", "ZMMLR", "ZMMMR"))
maizet <- maize %>% column_to_rownames(., var = "Sample_ID") %>% t() %>% as.data.frame() %>% 
  rownames_to_column(., var = "SNP_ID")
maize1 <- maizet[-c(1, 2), ]

teosinte <- filter(fang, Group %in% c("ZMPBA", "ZMPIL", "ZMPJA "))
teosintet <- teosinte %>% column_to_rownames(., var = "Sample_ID") %>% t() %>% as.data.frame() %>% 
  rownames_to_column(., var = "SNP_ID")
teosinte1 <- teosintet[-c(1, 2), ]
```

### Data Processing - select & merge 
* Select SNP_ID, Chromosome and Position from snp file 
* Merge with maize and teosinte file respectively

```{r select & merge}
join1 <- select(snp, SNP_ID, Chromosome, Position)
joinedmaize <- merge(join1, maize1, by = "SNP_ID", all = TRUE)
joinedteosinte <- merge(join1, teosinte1, by = "SNP_ID", all = TRUE)
```

#### Make directories for the outout files in the output folder 

```{r make directories}
dir.create("./maize_files")
dir.create("./teosinte_files")
```

### Data processing - mutate 
* Encode missing values, replace ? with - 
* Keep both versions (with and without question marks)

```{r mutate}
joinedmaizerep <- joinedmaize %>% 
 mutate_all(~ gsub("\\?", "-", .))

joinedteosinterep <- joinedteosinte %>% 
  mutate_all(~ gsub("\\?", "-", .))
```


### Function for separating files by chromosomes and then sapply
* Separate files by chromosome
* Sort by increasing or decreasing order
* Encode missing values with ? or -
* It gives warnings and long messages, so excluded but code looks like this:
({r function separating files, include = FALSE, message = FALSE, warning = FALSE})
separate_chr_files <- function(i) {
  maizeinc <- joinedmaize %>% filter(Chromosome == i) %>% arrange(as.numeric(Position))
  maizedec <- joinedmaizerep %>% filter(Chromosome == i) %>% arrange(desc(as.numeric(Position)))
  
  write_tsv(maizeinc, file.path("./maize_files/", paste("Maize_chr", i, "increasing.txt", sep = "_")))
  write_tsv(maizedec, file.path("./maize_files/", paste("Maize_chr", i, "decreasing.txt", sep = "_")))
  
  teoinc <- joinedteosinte %>% filter(Chromosome == i) %>% arrange(as.numeric(Position))
  teodec <- joinedteosinterep %>% filter(Chromosome == i) %>% arrange(desc(as.numeric(Position)))
  
  write_tsv(teoinc, file.path("./teosinte_files/", paste("Teosinte_chr", i, "_increasing.txt", sep = "_")))
  write_tsv(teodec, file.path("./teosinte_files/", paste("Teosinte_chr", i, "_decreasing.txt", sep = "_")))
 
}


sapply(1:10, separate_chr_files)


```{r function separating files, include = FALSE, message = FALSE, warning = FALSE}
separate_chr_files <- function(i) {
  maizeinc <- joinedmaize %>% filter(Chromosome == i) %>% arrange(as.numeric(Position))
  maizedec <- joinedmaizerep %>% filter(Chromosome == i) %>% arrange(desc(as.numeric(Position)))
  
  write_tsv(maizeinc, file.path("./maize_files/", paste("Maize_chr", i, "increasing.txt", sep = "_")))
  write_tsv(maizedec, file.path("./maize_files/", paste("Maize_chr", i, "decreasing.txt", sep = "_")))
  
  teoinc <- joinedteosinte %>% filter(Chromosome == i) %>% arrange(as.numeric(Position))
  teodec <- joinedteosinterep %>% filter(Chromosome == i) %>% arrange(desc(as.numeric(Position)))
  
  write_tsv(teoinc, file.path("./teosinte_files/", paste("Teosinte_chr", i, "_increasing.txt", sep = "_")))
  write_tsv(teodec, file.path("./teosinte_files/", paste("Teosinte_chr", i, "_decreasing.txt", sep = "_")))
 
}


sapply(1:10, separate_chr_files)

```



# Part 2 Visualization

### SNPs per chromosome for teosinte and maize 
* Number of SNPs per chromosome for teosinte and maize in barplots

```{r barplot, echo = FALSE}
joinedmaize$Chromosome <- factor(joinedmaize$Chromosome, levels = 1:10)
joinedmaizef <- joinedmaize %>%
  filter(!Chromosome %in% c("multiple", "unknown") & !is.na(Chromosome))

(plotmai <- ggplot(data = joinedmaizef) + 
  geom_bar(mapping = aes(x = Chromosome), fill = "darkgreen") +
  labs(x = "Chromosome", y = "SNP Count (Maize)", title = "Maize") +
  theme_minimal())


joinedteosinte$Chromosome <- factor(joinedteosinte$Chromosome, levels = 1:10)
joinedteosintef <- joinedteosinte %>%
  filter(!Chromosome %in% c("multiple", "unknown") & !is.na(Chromosome))

(plotteo <-ggplot(data = joinedteosintef) + 
  geom_bar(mapping = aes(x = Chromosome), fill = "yellow") +
  labs(x = "Chromosome", y = "SNP Count (Teosinte)", title = "Teosinte") +
  theme_minimal())

grid.arrange(plotmai, plotteo, ncol = 2)

```
### SNPs per chromosome for teosinte and maize 
* Density of SNPs per chromosome for teosinte and maize in barplots


```{r density, echo = FALSE}
(densitymai <- ggplot(joinedmaizef, aes(x = Chromosome, y = as.numeric(Position))) +
    geom_point(color = "darkgreen", alpha = 0.5) +
    labs(x = "Chromosome", y = "SNP Position", title = "Maize SNP Distribution") +
    theme_minimal())

(densityteo <- ggplot(joinedteosintef, aes(x = Chromosome, y = as.numeric(Position))) +
  geom_point(color = "yellow", alpha = 0.5) +
  labs(x = "Chromosome", y = "SNP Position", title = "Teosinte SNP Distribution") +
  theme_minimal())


grid.arrange(densitymai, densityteo, ncol = 2)

```

### Missing data and amount of heterozygosity 
* Create new column for homozyous/heterozygous or missing values (Characteristic) 
* Proportion of homozygous and heterozygous sites and missing data in stacked barplot

```{r stacked barplot}
fangnew <- fang %>% select(-JG_OTU) %>% 
  pivot_longer( -Sample_ID:-Group, names_to = "SNP_ID", values_to = "Basepairs")

fangnew <- fangnew %>% 
  mutate(Characteristic = ifelse(Basepairs %in% c("A/A","T/T","C/C","G/G"), "Homozygous", 
                               ifelse(Basepairs == "?/?", "Missing","Heterozygous")))


fang1 <- ggplot(fangnew, aes(x = Sample_ID, fill = Characteristic)) + geom_bar(position = "fill") + 
  theme_bw() + labs(x = "Sample ID", y = "Proportion")

fang2 <- ggplot(fangnew, aes(x = Group , fill = Characteristic)) + geom_bar(position = "fill") + 
  theme_bw() + theme(axis.text.x = element_text(angle = 90))+ labs(y = "Proportion")

grid.arrange(fang1, fang2, ncol = 2)
```


### Own Visualization
* Which chromosome is the shortest, which the longest based on position range in this dataset?

```{r own Visualization}
chromosomerange <- joinedmaizef %>%
  group_by(Chromosome) %>%
  summarise(position_range = max(as.numeric(Position), na.rm = TRUE) - min(as.numeric(Position), na.rm = TRUE)) %>%
  arrange(position_range)

shortest_chromosome <- chromosomerange %>% slice_min(position_range, n = 1)
longest_chromosome <- chromosomerange %>% slice_max(position_range, n = 1)

ggplot(chromosomerange, aes(x = factor(Chromosome), y = position_range)) +
  geom_bar(stat = "identity", fill = "lightblue") +
  labs(x = "Chromosome", y = "Position Range (Max - Min)", title = "Chromosome Position Ranges") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))

```

